ncm-dpmlfc/src/main/perl/DPMLFC/RuleBasedEditor.pm

```diff
@@ -0,0 +1,778 @@
+                                log => $self);
+  seek($fh, 0, SEEK_SET);
+
+  # Check that config file has an appropriate header
+  my $intro_pattern = "# This file is managed by Quattor";
+  my $intro = "# This file is managed by Quattor - DO NOT EDIT lines generated by Quattor";
+  $fh->add_or_replace_lines(qr/^$intro_pattern/,
+                            qr/^$intro$/,
+                            $intro."\n#\n",
+                            BEGINNING_OF_FILE,
+                           );
+  
+  # Loop over all config rule entries.
+  # Config rules are stored in a hash whose key is the variable to write
+  # and whose value is the rule itself.
+  # If the variable name start with a '-', this means that the matching configuration
+  # line must be commented out unconditionally.
+  # Each rule format is '[condition->]attribute:option_set[,option_set,...];line_fmt' where
+  #     condition: either a role that must be enabled or ALWAYS if the rule must be applied 
+  #                when 'always_rules_only' is true. A role is enabled if 'role_enabled' is
+  #                true in the corresponding option set.
+  #     option_set and attribute: attribute in option set that must be substituted
+  #     line_fmt: the format to use when building the line
+  # An empty rule is valid and means that the keyword part must be
+  # written as is, using the line_fmt specified.
+  
+  my $rule_id = 0;
+  foreach my $keyword (sort keys %{$config_rules}) {
+    my $rule = $config_rules->{$keyword};
+    $rule_id++;
+
+    # Initialize remove_if_undef flag according the default for this file
+    my $remove_if_undef = $parser_options->{remove_if_undef};
+
+    # Check if the keyword is prefixed by:
+    #     -  a '-': in this case the corresponding line must be unconditionally 
+    #               commented out if it is present
+    #     -  a '*': in this case the corresponding line must be commented out if
+    #               it is present and the option is undefined
+    my $comment_line = 0;
+    if ( $keyword =~ /^-/ ) {
+      $keyword =~ s/^-//;
+      $comment_line = 1;
+    } elsif ( $keyword =~ /^\?/ ) {
+      $keyword =~ s/^\?//;
+      $remove_if_undef = 1;
+      $self->debug(2,"$function_name: 'remove_if_undef' option set for the current rule");
+    }
+
+    # Split different elements of the rule
+    ($rule, my $line_fmt, my $value_fmt) = split /;/, $rule;
+    unless ( $line_fmt ) {
+      $line_fmt = LINE_FORMAT_DEFAULT;
+    }
+    my $value_opt;
+    if ( $value_fmt ) {
+      ($value_fmt, $value_opt) = split /:/, $value_fmt;
+    }else {
+      $value_fmt = LINE_VALUE_AS_IS;
+    }
+    unless ( defined($value_opt) ) {
+      $value_opt = LINE_VALUE_OPT_NONE;      
+    }
+
+    (my $condition, my $tmp) = split /->/, $rule;
+    if ( $tmp ) {
+      $rule = $tmp;
+    } else {
+      $condition = "";
+    }
+    $self->debug(1,"$function_name: processing rule ".$rule_id."(variable=>>>".$keyword.
```

Comment:
no need for the `".$something."`; also you can pass the debug message in multiple arguments, no need to make one long string